Add cmd get transactions by senderID and state filters - Closes #648

ishantiw opened the pull request:
### Description

Add command to get transactions based on `senderID`, `default`, `limit` and `offset` values.

Also related to #600 

Example,
`lisk transaction:get --sender-id=14523483941411160698L --state=unsigned`
`lisk transaction:get 7397859286435869218 --state=unprocessed`
`lisk transaction:get --limit=10 --offset=15`
`lisk transaction:get`

### Review checklist

* The PR resolves #648 
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

shuse2 commented on docs/transaction.md:
```diff
@@ -421,16 +421,27 @@ OPTIONS
   --[no-]pretty                     Prints JSON in pretty format rather than condensed. Has no effect if the output is
                                     set to table. You can change the default behaviour in your config.json file.
 
+  --senderId=senderId               Get transactions based by senderId which is sender's lisk address'.
```
Convention wise, I think it should be `sender-id`, as `second-passphrase`

shuse2 commented on src/commands/transaction/get.js:
```diff
@@ -13,13 +13,20 @@
  * Removal or modification of this copyright notice is prohibited.
  *
  */
+import { isArray } from 'util';
```
`Array.isArray`?

shuse2 commented on src/commands/transaction/get.js:
```diff
@@ -43,40 +50,110 @@ const queryNode = async (client, txnState, parameters) =>
 
-		if (txnState === 'unsigned') {
-			const results = await queryNode(client.node, txnState, req);
+		if (txnState) {
```
rather than this big if, if it's gonna return, maybe make scope smaller?

shuse2 commented on src/commands/transaction/get.js:
```diff
@@ -85,6 +162,17 @@ GetCommand.args = [
+	limit: flagParser.string({
+		description:
+			'Limits the returned transactions array by specified integer amount. Maximum is 100.',
+		default: '20',
```
from the documentation, it seems like default is `10`
https://lisk.io/documentation/lisk-core/user-guide/api/1-0#/Transactions/getTransactions

ishantiw replied:
@shuse2 I just checked the API and actually, it doesn't support sort by height. `Available values : amount:asc, amount:desc, fee:asc, fee:desc, type:asc, type:desc, timestamp:asc, timestamp:desc. Default value : amount:asc`
We can also keep it default that will sort it by amount or we keep it by height?

shuse2 commented on src/utils/query.js:
```diff
@@ -26,7 +26,7 @@ export const handleResponse = (endpoint, res, placeholder) => {
 			throw new Error(`No ${endpoint} found using specified parameters.`);
 		}
-		return res.data[0];
+		return res.data;
```
wouldn't this affect other queries?

ishantiw replied:
It was not returning a list of transactions from API so I had to change but I also tested all the commands which are using it and it works fine and also all the tests are passing.

shuse2 replied:
I think test passes because we stub the API response, but it changes the format ='(
```
> ./bin/run account:get 123L,456L                                                             
[
        {
                "0": {
                        "address": "123L",
                        "unconfirmedBalance": "6122216612",
                        "balance": "6122216612",
                        "publicKey": "",
                        "secondPublicKey": ""
                }
        },
        {
                "0": {
                        "address": "456L",
                        "unconfirmedBalance": "1000",
                        "balance": "1000",
                        "publicKey": "",
                        "secondPublicKey": ""
                }
        }
]
```

shuse2 commented on src/commands/transaction/get.js:
```diff
@@ -43,40 +50,110 @@ const queryNode = async (client, txnState, parameters) =>
+		if (!isArray(results)) {
+			return this.print(results);
+		}
+		const sortedResults = results.sort((a, b) => b.height - a.height);
```
In stead of doing `sort` in the code, probably better to use API sort field

ishantiw replied:
@shuse2 I just checked the API and actually, it doesn't support sort by height. `Available values : amount:asc, amount:desc, fee:asc, fee:desc, type:asc, type:desc, timestamp:asc, timestamp:desc. Default value : amount:asc`
We can also keep it default that will sort it by amount or we keep it by height?

mitsujutsu commented on src/commands/transaction/get.js:
```diff
@@ -43,40 +50,110 @@ const queryNode = async (client, txnState, parameters) =>
+				return this.print(results);
+			}
+
+			if (senderAddress) {
```
should show some error if provided senderId does not exist or invalid. currently doesn't seem to throw.

ishantiw commented:
Addressed above concerns.

Considering all the arguments and flags, below are some examples for testing

`transaction:get 10041151099734832021`
`transaction:get 10041151099734832021,1260076503909567890`
`transaction:get 10041151099734832021,1260076503909567890 --state=unprocessed`
`transaction:get --state=unsigned --sender-id=1813095620424213569L`
`transaction:get 10041151099734832021 --state=unsigned --sender-id=1813095620424213569L`
`transaction:get --sender-id=1813095620424213569L`
`transaction:get --limit=10 --sort=amount:desc`
`transaction:get --limit=10 --offset=5`

shuse2 commented on src/commands/transaction/get.js:
```diff
@@ -43,40 +59,151 @@ const queryNode = async (client, txnState, parameters) =>
-			const results = await queryNode(client.node, txnState, req);
+
+		if (txnState && ids) {
+			const reqTransactionIDs = ids.map(id => ({
```
why not `reqTransactionIds` where other places are using `Id`?

shuse2 commented on test/commands/transaction/get.test.js:
```diff
@@ -163,6 +158,109 @@ describe('transaction:get', () => {
 					return expect(printMethodStub).to.be.calledWithExactly(queryResult);
 				},
 			);
+		describe('transaction: get --sender-id', () => {
```
prefer `transaction:get` without space

ishantiw commented on src/commands/transaction/get.js:
```diff
@@ -43,40 +59,151 @@ const queryNode = async (client, txnState, parameters) =>
+		};
 		const results = await query(client, 'transactions', req);
 
+		if (!Array.isArray(results)) {
```
Now after fixing query handler it's not needed

ishantiw commented on test/commands/transaction/get.test.js:
```diff
@@ -163,6 +158,109 @@ describe('transaction:get', () => {
+					sandbox.stub().resolves({ message: 'No transactions found.' }),
+				)
+				.command(['transaction:get', '--offset=10'])
+				.it('should return a message with no transactions found.', () => {
```
Actually, this test is only to check offset by assuming that there are no results after 10 txns as configured in the stub. So, it should return results if a node has more than 10 txns.

ishantiw commented on test/commands/transaction/get.test.js:
```diff
@@ -194,60 +292,94 @@ describe('transaction:get', () => {
 		};
 
-		describe('transaction: get transactions --state-unprocessed', () => {
+		setupTest()
```
For all queries related to getting transactions by a state is not using the general query handler but its own handler so it's not used for these test suits.

mitsujutsu replied:
Ok I think we should also test for custom queryNode handler in the same way

ishantiw replied:
@mitsujutsu discussed with @shuse2, I will add it in testing

ishantiw commented:
@shuse2 fixed query handler, updated it to handle arrays more than 1 records. This time I tested it with the right build 😅 but please test it again on your end also.

@mitsujutsu addressed 👍

mitsujutsu commented on test/commands/transaction/get.test.js:
```diff
@@ -84,18 +77,20 @@ describe('transaction:get', () => {
 			.stub(queryHandler, 'query', sandbox.stub().resolves(queryResult))
 			.command(['transaction:get', transactionIds.join(',')])
-			.it('should get two transactions’ info and display as an array', () => {
+			.it('should get two transactions’ info and display as an array.', () => {
```
`transactions’` --> `transaction's`
